std/range/package.d

```diff
@@ -1574,18 +1574,20 @@ Returns:
     alias of that range's type.
  */
 auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length > 2
-        && is(typeof(choose(true, rs[0], rs[1])))
-        && is(typeof(chooseAmong(0, rs[1 .. $]))))
+if (Ranges.length >= 2 && allSatisfy!(isInputRange, staticMap!(Unqual, Ranges)))
 {
-    return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
-}
-
-/// ditto
-auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
-{
-    return choose(index == 0, rs[0], rs[1]);
+    static if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
+    {
+        return choose(index == 0, rs[0], rs[1]);
+    }
+    else static if (is(typeof(choose(true, rs[0], rs[1]))) && is(typeof(chooseAmong(0, rs[1 .. $]))))
+    {
+        return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
+    }
+    else
+    {
+        static assert(0, "The ranges have no common type.");
```

Comment:
This is new behavior - now `chooseAmong` accepts wrong arguments and errs in the implementation. That is acceptable, but the current prevalent approach is to eliminate unacceptable arguments in the constraint. To do so:

```
auto chooseAmong(Ranges...)(size_t index, Ranges rs)
if (Ranges.length >= 2
    && allSatisfy!(isInputRange, staticMap!(Unqual, Ranges))
    && !is(CommonType!(staticMap!(ElementType, Ranges)) == void))
{
    ...
}
```

It's also nicely readable - there have got to be two or more arguments, all of which must be input ranges, which must have some common element type.